### PR TITLE
Resizing a fullscreen element doesn't always resize children to the correct size

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### d7f28b8f18d6f400feac62dc7054759e91014765
<pre>
Resizing a fullscreen element doesn&apos;t always resize children to the correct size
<a href="https://bugs.webkit.org/show_bug.cgi?id=258847">https://bugs.webkit.org/show_bug.cgi?id=258847</a>
&lt;radar://110880116&gt;

Reviewed by Tim Horton.

During a window resize, [_contentView bounds] will not be up to date during the precommit
handler so it is incorrect to intersect it with the web view&apos;s bounds.

Additionally it makes fullscreen resizing look much smoother.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateVisibleContentRects]):
Skip intersection if we just resized the WKWebView.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/WKWebViewResize.mm: Added.
(-[NSObject swizzled_didUpdateVisibleRect:unobscuredRect:contentInsets:unobscuredRectInScrollViewCoordinates:obscuredInsets:unobscuredSafeAreaInsets:inputViewBounds:scale:minimumScale:viewStability:enclosedInScrollableAncestorView:sendEvenIfUnchanged:]):
(setupWKContentViewSwizzle):
(operator==):
Required for EXPECT_EQ on CGRect.

(webViewHasExpectedBounds):
(TEST):
Add API test which ensures a content view with no other insets will have an
unobscuredRect which matches the enclosing WKWebView&apos;s bounds after setFrame: is called.

Tests both resizing the web view larger and smaller.

Canonical link: <a href="https://commits.webkit.org/266257@main">https://commits.webkit.org/266257@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3e94b6df9ed530c0a40b10e2d0d97d0482187009

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15065 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12689 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13407 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13668 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15370 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14151 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11270 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15696 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11439 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12025 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19077 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12514 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12192 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15405 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10554 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11966 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3259 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12537 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderVideo.cpp [hunk 1]
```diff
@@ -294,6 +295,15 @@ void RenderVideo::updateFromElement()
     updatePlayer();
 }
 
+bool RenderVideo::inFullscreenElementOrFullscreenVideo() const
```
Comment (eric-carlson) [inline Source/WebCore/rendering/RenderVideo.cpp]: Nit: "VideoFullscreen" is used more commonly used than "FullscreenVideo", so maybe `inFullscreenElementOrVideoFullscreen`?

Comment (mwyrzykowski) [reply Source/WebCore/rendering/RenderVideo.cpp]: Sounds good, I'll change that.

File: Source/WebCore/dom/Document.cpp [hunk 2]
```diff
@@ -2126,7 +2126,11 @@ void Document::resolveStyle(ResolveStyleType type)
 
         m_inStyleRecalc = true;
 
-        if (m_needsFullStyleRebuild)
+        bool needsFullStyleRebuild = m_needsFullStyleRebuild;
+#if ENABLE(FULLSCREEN_API)
+        needsFullStyleRebuild = needsFullStyleRebuild || m_fullscreenManager->isResizing();
```
Comment (smfr) [inline Source/WebCore/dom/Document.cpp]: Seems odd that stye resolves needs to know anything about fullscreen. Why doesn't window resize have the same issue? I think we should debug further.

**Review by smfr — CHANGES_REQUESTED** [review]

File: Source/WebCore/dom/Document.cpp (hunk 2 above)
Comment (mwyrzykowski) [reply Source/WebCore/dom/Document.cpp]: I can also fix this by calling `scheduleFullStyleRebuild` in a few places but it is a slightly larger change. Do you think that would be reasonable? I can try that instead.

Comment (nt1m) [reply Source/WebCore/dom/Document.cpp]: Fwiw, full style rebuilds were originally there because we had the UA styles matching the `:-webkit-full-screen-ancestor` pseudo class requiring _all_ the ancestors of the fullscreen element to have their styles recomputed.

It is in theory no longer required with the new fullscreen API (for which I did remove full style rebuilds in some places but not all). The remaining ones avoid dirty renderer assertions when pushing to top layer (which we should arguably fix and then remove the full style rebuilds).

Adding more full style rebuilds isn't really the answer here. Fine grained PseudoClassChangeInvalidation should do its job here.

File: Source/WebCore/rendering/RenderVideo.cpp [hunk 3]
```diff
@@ -29,6 +29,7 @@
 #include "RenderVideo.h"
```
Comment (mwyrzykowski) [inline Source/WebCore/rendering/RenderVideo.cpp]: The changes in RenderVideo.h/cpp are to fix the same issue 265672@main fixed but for video playing in a fullscreen element.

They can be broken out into a separate patch if preferred, the issues are independent.

File: Source/WebCore/page/LocalFrameView.cpp [hunk 4]
```diff
@@ -3292,6 +3292,8 @@ void LocalFrameView::unobscuredContentSizeChanged()
 {
 #if PLATFORM(IOS_FAMILY)
     updateTiledBackingAdaptiveSizing();
+    if (auto* document = m_frame->document())
+        document->scheduleFullStyleRebuild();
```
Comment (mwyrzykowski) [inline Source/WebCore/page/LocalFrameView.cpp]: This might not be sufficient. Layout is generally but not always occurring after a full style rebuild.

Maybe I want `layoutContext().layout();` instead?

**Comment by mwyrzykowski** [comment]
@hortont424 @smfr I added a test covering this behavior, please let me know if you have any further feedback or would like any changes to be made.

I also cloned this issue to track if skipping the intersection is the correct thing to do in all cases when the window is resized. We were effectively already doing something similar when the window was resized smaller, since the content view bounds were still the larger size, so I don't think this change introduces an issue which was not previously present.

**Review by hortont424 — APPROVED** [review]
